[UnifiedPDF] Selection painting requests should be optimized by specifying dirty rects

aprotyas opened the pull request:
#### 76578d58608418ccb9ed51f0a5363763a29d4e18
<pre>
[UnifiedPDF] Selection painting requests should be optimized by specifying dirty rects
<a href="https://bugs.webkit.org/show_bug.cgi?id=268927">https://bugs.webkit.org/show_bug.cgi?id=268927</a>
<a href="https://rdar.apple.com/122481810">rdar://122481810</a>

Reviewed by Tim Horton.

Currently, we repaint a lot more than is necessary for selection
changes. This commit optimizes our repaint requests on selection changes
by only asking for the rects corresponding to the bounds of the current
selection (and the old selection) to be repainted.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setNeedsRepaintInDocumentRects):
(WebKit::UnifiedPDFPlugin::windowActivityDidChange):
(WebKit::UnifiedPDFPlugin::selectionBoundsAcrossDocument const):
(WebKit::UnifiedPDFPlugin::repaintOnSelectionActiveStateChangeIfNeeded):
(WebKit::UnifiedPDFPlugin::protectedCurrentSelection const):
(WebKit::UnifiedPDFPlugin::setCurrentSelection):

Canonical link: <a href="https://commits.webkit.org/275359@main">https://commits.webkit.org/275359@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9355bda2fca30c7dd4b198ad7569cb367250efcb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20656 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44022 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44211 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/37734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23771 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17986 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42216 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17571 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35855 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45588 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37817 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37191 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40941 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/16457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13504 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/39357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18076 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9324 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/18132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -250,7 +250,8 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     void stopTrackingSelection();
     void setCurrentSelection(RetainPtr<PDFSelection>&&);
-    void repaintOnSelectionActiveStateChangeIfNeeded(ActiveStateChangeReason);
+    Vector<WebCore::FloatRect> selectionBoundsAcrossDocument(const RetainPtr<PDFSelection>&) const;
```
We don't normally pass smart pointers by const reference. Odds are you should be passing a `const PDFSelection*`

aprotyas replied:
Changed to `const PDFSelection *` in https://github.com/WebKit/WebKit/commit/9355bda2fca30c7dd4b198ad7569cb367250efcb

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -292,6 +293,12 @@ static String mutationObserverNotificationString()
 
+void UnifiedPDFPlugin::setNeedsRepaintInDocumentRects(OptionSet<RepaintRequirement> repaintRequirements, const Vector<FloatRect>& rectsInDocumentCoordinates)
+{
+    for (const auto& rectInDocumentCoordinates : rectsInDocumentCoordinates)
```
`const ` is unnecessary and usually omitted in such cases.

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/9355bda2fca30c7dd4b198ad7569cb367250efcb

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2548,14 +2567,20 @@ static FloatRect computeMarqueeSelectionRect(const WebCore::FloatPoint& point1,
     }
 
-    m_contentsLayer->setNeedsDisplay();
+    auto selectionDocumentRectsToRepaint = selectionBoundsAcrossDocument(m_currentSelection);
```
You should have a RetainPtr to m_currentSelection *on the stack* before passing it as parameter:
```
selectionBoundsAcrossDocument(RetainPtr { m_currentSelection }.get());
```
or introduce a member function named `protectedCurrentSelection()` which returns a `RetainPtr` and use it here:
```
selectionBoundsAcrossDocument(protectedCurrentSelection().get());
```

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/9355bda2fca30c7dd4b198ad7569cb367250efcb

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2565,10 +2590,12 @@ static FloatRect computeMarqueeSelectionRect(const WebCore::FloatPoint& point1,
 
 void UnifiedPDFPlugin::setCurrentSelection(RetainPtr<PDFSelection>&& selection)
 {
+    auto staleSelectionDocumentRectsToRepaint = selectionBoundsAcrossDocument(m_currentSelection);
```
ditto

aprotyas replied:
Fixed!

aprotyas replied:
https://github.com/WebKit/WebKit/commit/9355bda2fca30c7dd4b198ad7569cb367250efcb

hortont424 reviewed (approved):
r+ if you fix all of Chris' comments :P

aprotyas committed:
[UnifiedPDF] Selection painting requests should be optimized by specifying dirty rects
https://bugs.webkit.org/show_bug.cgi?id=268927
rdar://122481810

Reviewed by Tim Horton.

Currently, we repaint a lot more than is necessary for selection
changes. This commit optimizes our repaint requests on selection changes
by only asking for the rects corresponding to the bounds of the current
selection (and the old selection) to be repainted.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setNeedsRepaintInDocumentRects):
(WebKit::UnifiedPDFPlugin::windowActivityDidChange):
(WebKit::UnifiedPDFPlugin::selectionBoundsAcrossDocument const):
(WebKit::UnifiedPDFPlugin::repaintOnSelectionActiveStateChangeIfNeeded):
(WebKit::UnifiedPDFPlugin::protectedCurrentSelection const):
(WebKit::UnifiedPDFPlugin::setCurrentSelection):

Canonical link: https://commits.webkit.org/275359@main
